pkg/api/api.go

```diff
@@ -515,7 +521,22 @@ func (hs *HTTPServer) registerRoutes() {
 
 		// metrics
 		// DataSource w/ expressions
-		apiRoute.Post("/ds/query", requestmeta.SetSLOGroup(requestmeta.SLOGroupHighSlow), authorize(ac.EvalPermission(datasources.ActionQuery)), routing.Wrap(hs.QueryMetricsV2))
+		dsQueryHandler := routing.Wrap(hs.QueryMetricsV2)
+		if hs.Features.IsEnabledGlobally(featuremgmt.FlagKubernetesQueryServiceRewrite) {
+			// DEV ONLY FEATURE FLAG!
+			// rewrite requests from /ds/query to the new query service
+			namespaceMapper := request.GetNamespaceMapper(hs.Cfg)
+			dsQueryHandler = func(w http.ResponseWriter, r *http.Request) {
+				user, err := appcontext.User(r.Context())
+				if err != nil || user == nil {
+					errhttp.Write(r.Context(), fmt.Errorf("no user"), w)
+					return
+				}
+				r.URL.Path = "/apis/query.grafana.app/v0alpha1/namespaces/" + namespaceMapper(user.OrgID) + "/query"
+				hs.clientConfigProvider.DirectlyServeHTTP(w, r)
+			}
+		}
+		apiRoute.Post("/ds/query", requestmeta.SetSLOGroup(requestmeta.SLOGroupHighSlow), authorize(ac.EvalPermission(datasources.ActionQuery)), dsQueryHandler)
```

Comment:
Just meant that we don't normally put any handler code/implementation  in this file api.go why it could move to metrics.go for now?

Reply:
I'm not sure how to handle the `routing.Wrap(...)` part 🤔  -- the replacement is unable to return a `response.Response`, it is simply a url rewrite

Reply:
Good point/question :smile: Maybe something like below, otherwise keep as is for now and we can figure out later

```go
// api.go
apiRoute.Post("/ds/query", requestmeta.SetSLOGroup(requestmeta.SLOGroupHighSlow), authorize(ac.EvalPermission(datasources.ActionQuery)), hs.DSQueryEndpoint)

// metrics.go
func (hs *HTTPServer) DSQueryEndpoint() web.Handler {
	if hs.Features.IsEnabledGlobally(featuremgmt.FlagKubernetesQueryServiceRewrite) {
		// DEV ONLY FEATURE FLAG!
		// rewrite requests from /ds/query to the new query service
		namespaceMapper := request.GetNamespaceMapper(hs.Cfg)
		return func(w http.ResponseWriter, r *http.Request) {
			user, err := appcontext.User(r.Context())
			if err != nil || user == nil {
				errhttp.Write(r.Context(), fmt.Errorf("no user"), w)
				return
			}
			r.URL.Path = "/apis/query.grafana.app/v0alpha1/namespaces/" + namespaceMapper(user.OrgID) + "/query"
			hs.clientConfigProvider.DirectlyServeHTTP(w, r)
		}
	}

	return routing.Wrap(hs.QueryMetricsV2)
}
```